include/unix/osd.h

```diff
@@ -51,6 +51,43 @@
 #define OFI_UNUSED UNREFERENCED_PARAMETER
 #endif
 
+/* Atomic memory access macro */
+#define ofi_atomic_inc32(ptr)			\
+	__sync_fetch_and_add((ptr), (uint32_t)1)
+
+#define ofi_atomic_inc64(ptr)			\
+	__sync_fetch_and_add((ptr), (uint64_t)1)
+
+#define ofi_atomic_dec32(ptr)			\
+	__sync_fetch_and_sub((ptr), (uint32_t)1)
+
+#define ofi_atomic_dec64(ptr)			\
+	__sync_fetch_and_sub((ptr), (uint64_t)1)
+
+#define ofi_atomic_add32(ptr, value)			\
+	__sync_fetch_and_add((ptr), (uint32_t)(value))
+
+#define ofi_atomic_add64(ptr, value)			\
+	__sync_fetch_and_add((ptr), (uint64_t)(value))
+
+#define ofi_atomic_sub32(ptr, value)			\
+	__sync_fetch_and_sub((ptr), (uint32_t)(value))
+
+#define ofi_atomic_sub64(ptr, value)			\
+	__sync_fetch_and_sub((ptr), (uint64_t)(value))
+
+#define ofi_val_compare_and_swap_ptr(ptr, ptr1, ptr2)		\
+	__sync_val_compare_and_swap((ptr), (ptr1), (ptr2))
+
+#define ofi_val_compare_and_swap32(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap64(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap128(ptr, oldval, newval, compare)	\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
```

Comment:
Please convert all of these into static inline calls, so that we have proper type checking and know what type of value each function returns.  We need to verify that all implementations return the same value.  For example always return the value after an update, or always return the value prior to the update.